crates/iceberg/src/writer/file_writer/location_generator.rs

```diff
@@ -0,0 +1,251 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! The module contains the location generator for file writer.
+
+use std::str::FromStr;
+use std::sync::atomic::AtomicUsize;
+use uuid::Uuid;
+
+use crate::spec::{DataFileFormat, TableMetadata};
+use crate::{Error, ErrorKind, Result};
+
+/// LocationGenerator will generate a file location for the writer.
+pub trait LocationGenerator: Send + Sync + 'static {
+    /// Generate a related file location for the writer.
```

Comment:
> This comment is a little confusing, is the result a full path or related to table location?

A rel path(include data dir)for data file. e.g. data/xxx.parquet. I refine it later.

 > Is this similar to LocationProvider in java API?

No. I find that the interface is not similar. `LocationProvider` seems a higher level abstract. It return a fully-qualified data file location for the given filename so that we also need to provide a file name.

The `LocationGenerator` will generate a fully-qualified file name.🤔

Reply:
I feel the `LocationProvider` interface make more sense to me: it asks user to provide filename and partition value. We can't guess file suffix from table property, I think to refine the interface, we should ask user to provide things such as `DataFileFormat`, partition value.

Reply:
LGTM. How about keeping the same interface as `LocationProvider`? We can combine the `DataFileFormat` outside to keep the interface of `LocationProvider` simple.

Reply:
Sounds reasonable to me.

Reply:
For now, the writer doesn't need the LocationProvider. I will implement this `LocationProvider` in another PR.